### PR TITLE
Added >=0 check on delta

### LINKED ISSUE #16254: Date Shows
## Description

Last Modified date shows "next year" for new notebooks and existing notebooks soon after being saved, especially while
under a heavy processing load.

## Reproduce

1. Create a new notebook:

image

2. Click on File > Save Notebook/Press CTRL+S/Click Save Button and save the notebook

image

3. Add some code that will take a long time to process, run it, and then click save:

before

after clicking save

Diagnosed and narrowed the issue down to `Time.formatHuman` function in
`jupyter/jupyterlab/packages/coreutils/src/time.ts`. If you set a breakpoint after `delta` is calculated it is sometimes
a positive value. My guess is this is due to some asynchronous execution resulting in the save time appearing before the
Date.now() time.

image

With `delta` (`s`) being a positive value, the result of `Math.ceil(delta / unit.milliseconds)` for any of the units
will be 1. The first unit in the list is `{ name: 'years', milliseconds: 365 * 24 * 60 * 60 * 1000 }` and thus with
amount being non-zero `return formatter.format(amount, unit.name);` on line 57 has the values 1, "year" and is formatted
to "next year".

A simple check to see if s is positive and return `formatter.format(0, 'seconds');` if it is will resolve the issue and
a PR is to follow submitted by me; just wanted to get the issue documented first.

## Expected behavior

Expect it to show "0 seconds" or "now" or whatever makes sense based on notebook language settings.

## Context

- Operating System and version: Debian GNU/Linux 12 (bookworm)
- Browser and version:
  - Firefox 125.0.2 (64-bit)
  - Edge 124.0.2478.51 (Official build) (64-bit)
- JupyterLab version: 4.1.5

<details><summary>Troubleshoot Output</summary>
<pre>
$PATH:
        /home/nuvious/EN.695.737.81.SP24-ArtificialIntelligenceForAssuredAutonomy/.venv/bin
        /home/nuvious/.vscode-server/cli/servers/Stable-e170252f762678dec6ca2cc69aba1570769a5d39/server/bin/remote-cli
        /usr/local/bin
        /usr/bin
        /bin
        /usr/local/games
        /usr/games

sys.path:
        /home/nuvious/EN.695.737.81.SP24-ArtificialIntelligenceForAssuredAutonomy/.venv/bin
        /usr/lib/python311.zip
        /usr/lib/python3.11
        /usr/lib/python3.11/lib-dynload
        /home/nuvious/EN.695.737.81.SP24-ArtificialIntelligenceForAssuredAutonomy/.venv/lib/python3.11/site-packages

sys.executable:
        /home/nuvious/EN.695.737.81.SP24-ArtificialIntelligenceForAssuredAutonomy/.venv/bin/python3

sys.version:
        3.11.2 (main, Mar 13 2023, 12:18:29) [GCC 12.2.0]

platform.platform():
        Linux-6.1.0-18-amd64-x86_64-with-glibc2.36

which -a jupyter:
        /home/nuvious/EN.695.737.81.SP24-ArtificialIntelligenceForAssuredAutonomy/.venv/bin/jupyter

pip list:
        Package                      Version
        ---------------------------- --------------
        absl-py                      2.1.0
        amalgam-lang                 10.0.0
        anyio                        4.3.0
        argon2-cffi                  23.1.0
        argon2-cffi-bindings         21.2.0
        arrow                        1.3.0
        asttokens                    2.4.1
        astunparse                   1.6.3
        async-lru                    2.0.4
        attrs                        23.2.0
        Babel                        2.14.0
        bayesian-optimization        1.4.3
        beautifulsoup4               4.12.3
        bleach                       6.1.0
        certifi                      2024.2.2
        cffi                         1.16.0
        charset-normalizer           3.3.2
        cloudpickle                  3.0.0
        colorama                     0.4.6
        comm                         0.2.2
        contourpy                    1.2.0
        cryptography                 42.0.5
        cycler                       0.12.1
        debugpy                      1.8.1
        decorator                    5.1.1
        defusedxml                   0.7.1
        deprecation                  2.1.0
        executing                    2.0.1
        Faker                        24.3.0
        fastjsonschema               2.19.1
        flatbuffers                  24.3.7
        fonttools                    4.50.0
        fqdn                         1.5.1
        gast                         0.5.4
        google-pasta                 0.2.0
        grpcio                       1.62.1
        h11                          0.14.0
        h5py                         3.10.0
        howso-engine                 19.0.0
        howso-openapi-client         32.0.0
        httpcore                     1.0.4
        httpx                        0.27.0
        humanize                     4.9.0
        idna                         3.6
        imageio                      2.34.1
        ipykernel                    6.29.3
        ipython                      8.22.2
        ipywidgets                   8.1.2
        isoduration                  20.11.0
        jedi                         0.19.1
        Jinja2                       3.1.3
        joblib                       1.3.2
        json5                        0.9.24
        jsonpointer                  2.4
        jsonschema                   4.21.1
        jsonschema-specifications    2023.12.1
        jupyter                      1.0.0
        jupyter_client               8.6.1
        jupyter-console              6.6.3
        jupyter_core                 5.7.2
        jupyter-events               0.10.0
        jupyter-lsp                  2.2.4
        jupyter_server               2.13.0
        jupyter_server_terminals     0.5.3
        jupyterlab                   4.1.5
        jupyterlab_pygments          0.3.0
        jupyterlab_server            2.25.4
        jupyterlab_widgets           3.0.10
        keras                        3.1.1
        kiwisolver                   1.4.5
        lazy_loader                  0.4
        libclang                     18.1.1
        lime                         0.2.0.1
        llvmlite                     0.42.0
        Markdown                     3.6
        markdown-it-py               3.0.0
        MarkupSafe                   2.1.5
        matplotlib                   3.8.3
        matplotlib-inline            0.1.6
        mdurl                        0.1.2
        mistune                      3.0.2
        ml-dtypes                    0.3.2
        mmh3                         4.1.0
        namex                        0.0.7
        nbclient                     0.10.0
        nbconvert                    7.16.3
        nbformat                     5.10.3
        nest-asyncio                 1.6.0
        networkx                     3.3
        notebook                     7.1.2
        notebook_shim                0.2.4
        numba                        0.59.1
        numpy                        1.26.4
        opt-einsum                   3.3.0
        optree                       0.10.0
        overrides                    7.7.0
        packaging                    24.0
        pandas                       2.2.1
        pandocfilters                1.5.1
        parso                        0.8.3
        pexpect                      4.9.0
        pillow                       10.2.0
        pip                          23.0.1
        platformdirs                 4.2.0
        prometheus_client            0.20.0
        prompt-toolkit               3.0.43
        protobuf                     4.25.3
        psutil                       5.9.8
        ptyprocess                   0.7.0
        pure-eval                    0.2.2
        pycparser                    2.21
        pygad                        3.3.1
        Pygments                     2.17.2
        PyJWT                        2.8.0
        pyparsing                    3.1.2
        python-dateutil              2.9.0.post0
        python-json-logger           2.0.7
        pytz                         2024.1
        PyYAML                       6.0.1
        pyzmq                        25.1.2
        qtconsole                    5.5.1
        QtPy                         2.4.1
        referencing                  0.34.0
        requests                     2.31.0
        rfc3339-validator            0.1.4
        rfc3986-validator            0.1.1
        rich                         13.7.1
        rpds-py                      0.18.0
        scikit-image                 0.23.2
        scikit-learn                 1.4.1.post1
        scipy                        1.12.0
        semantic-version             2.10.0
        Send2Trash                   1.8.2
        setuptools                   66.1.1
        shap                         0.45.0
        six                          1.16.0
        slicer                       0.0.7
        sniffio                      1.3.1
        soupsieve                    2.5
        stack-data                   0.6.3
        tensorboard                  2.16.2
        tensorboard-data-server      0.7.2
        tensorflow                   2.16.1
        tensorflow-io-gcs-filesystem 0.36.0
        termcolor                    2.4.0
        terminado                    0.18.1
        threadpoolctl                3.4.0
        tifffile                     2024.4.24
        tinycss2                     1.2.1
        tornado                      6.4
        tqdm                         4.66.2
        traitlets                    5.14.2
        types-python-dateutil        2.9.0.20240316
        typing_extensions            4.10.0
        tzdata                       2024.1
        uri-template                 1.3.0
        urllib3                      2.2.1
        wcwidth                      0.2.13
        webcolors                    1.13
        webencodings                 0.5.1
        websocket-client             1.7.0
        Werkzeug                     3.0.1
        wheel                        0.43.0
        widgetsnbextension           4.0.10
        wrapt                        1.16.0
</pre>
</details>

<details><summary>Command Line Output</summary>
<pre>
N/A - Does not throw an exception and source of issue was isolated with javascript/typescript debugging and static
analysis.
</pre>
</details>

<details><summary>Browser Output</summary>

<pre>
See screenshot for specific issue found in TypeScript
</pre>
</details>

### PR DESCRIPTION
**Pull request by nuvious**
## References

Alleviates #16254

## Code changes

Added check on delta to see if it is positive. If it is, return 0 for value and 'seconds' for the unit per documentation below and convention demonstrated at line [48](https://github.com/jupyterlab/jupyterlab/blob/main/packages/coreutils/src/time.ts#L48)/52  (current/proposed time.ts implementation).

Without this check if delta is greater than 0, `Math.ceil(delta / unit.milliseconds)` will compute to 1 with the units being "years" in the first check resulting in the behavior outlined in #16254

## User-facing changes

User will se 'now' or '0 seconds'. Had trouble building/installing jupterlab from source in a fresh venv and still trying to so I can provide manual testing validation

## Backwards-incompatible changes

N/A

### CODE REVIEW

**Comment by JasonWeill** [comment]
@nuvious Thank you so much for opening this issue and pull request! It is possible, using commands like `touch`, to manually set a date in the future for a particular file. It looks like your case covers a situation where the delta is only slightly larger than 0 (by a second or so) and is being misreported as 1 year from now. There are other ways to handle this:

1. Treat files with a modified date within a few seconds of now as "now" (common on many hosted software products)
2. Revising the logic around the `ceil` function to handle cases where the file's datestamp is in the future.

I don't think that treating _all_ future-dated files as being modified "now" is the best approach to address #16254.

It would also be good to improve the unit test coverage for future-dated files.

**Comment by nuvious** [comment]
I think it is reasonable to just treat anything resolving to the future as "now". This is a narrow edge case and the value is "Last Modified" which implies a context in the past. Last modified should never be in the future because that doesn't make sense in any context. The >=0 check acts as an input validation for the ceiling function for this edge case. That's all that's really needed here is some input validation.

I can try my hands at the unit tests but need to familiarize myself with the backend a bit. Will look into that when I get a chance. Also have started using the docker/start.sh environment builder as well to do manual testing and should be able to provide that info in the near future.

**Comment by krassowski** [comment]
Any updates @nuvious?

**Comment by nuvious** [comment]
Hey @krassowski, looking at it again today between work and school stuff. Got the development environment up today and did some manual tests to evaluate if the fix is working. Going to try to get into the test suite stuff today; not sure why my changes would result in the above failures, but I can dig a little deeper as I familiarize myself with the test pipelines.

File: packages/coreutils/src/time.ts [hunk 1]
```diff
@@ -36,6 +36,10 @@ export namespace Time {
       style: format
     });
     const delta = new Date(value).getTime() - Date.now();
+    // Delta may be positive due to async execution; default to 0 seconds if it is
+    if (delta >= 0) {
+      return formatter.format(0, "seconds");
```
Comment (JasonWeill) [inline packages/coreutils/src/time.ts:41]:
```suggestion
      return formatter.format(0, 'seconds');
```

This would fix the lint error reported in the pipeline.

**Comment by nuvious** [comment]
@krassowski, dusted this off after a comment got left on the referenced issue. Got the devcontainer up and running and tested my changes after a merge with the upstream main. Going to test it again locally some more but just letting the CICD run to see what fails. I noticed a screenshot failed the last run and I'm going to look into that one specifically if it fails. May need some guidance on updating the screenshot for the test if the referenced issue it the reason the test failed.

**Comment by nuvious** [comment]
P.S. I had to add `git config --global --add safe.directory /workspaces/jupyterlab` to the postCreateCommand in the devcontainer definition before the micromamba call as well as as the first command run in the final `RUN` command in the Dockerfile to get the dev container to work. May be missing something about local dev using the devcontainer, but the build process was erroring out on untrusted directories on some git command calls.

**Comment by nuvious** [comment]
@JasonWeill / @krassowski, have a few alternative proposals if the simple >=0 check is still not desired.

First is to consider anything within 5 seconds to be defaulted to 0 seconds or 'now':

```typescript
    // Delta may be positive due to async execution; consider anything less than 5 seconds to be 'now'
    if (Math.abs(delta) < 5) {
      return formatter.format(0, 'seconds');
    }
```

Second is to adjust the calculation of amount to be execute a `Math.min` on the amount calculation so if it's positive it's set to 0 as an alternative to the if-statement based solution already in the queue.

```typescript
    const amount = Math.floor(Math.ceil(delta / unit.milliseconds), 0);
```

Would prefer the later solution over the former since any delta that projects into the future is likely an error. It's possible someone may have their system clock set wrong, but in that context the delta since last save is completely unreliable at that point. Think it's elegant to just treat any positive delta as 'now' in the context of displaying how long since the last edit to the user.

**Comment by nuvious** [comment]
Just following up on this I'm still unsure of how to test this. Is there a chance we could close this out and then open up a different "find a better solution to" issue? I'm unfamiliar with how I could force this discrepancy in testing. Still standing by for any guidance on the above questions when you have time.

**Comment by linlol** [comment]
@krassowski can we merge to have the issue sorted 

Seems like this case is hard to be covered in ut since it cannot be stably reproduced 😭

well yes probably it's worthwhile to double check...

**Comment by linlol** [comment]
@krassowski @JasonWeill Sorry for pinging you here, but may I know your idea about this

would you like a unit tests only or would you want a more state of the art and permaent fix otherthan simply a error-covered fix

**Comment by krassowski** [comment]
I see this issue pop up consistently on Binder. I think there are three things that need to be done:
- we should add a `console.warning` when delta is more than one minute suggesting that the server time may be out of sync
- we should investigate why Binder is running on a server with clock offset by ~30 seconds (which is a big difference IMO)
- we should add a unit test case for the changed function

image

**Comment by krassowski** [comment]
I think this masks the problem and will make the timing issues harder to diagnose. I think we should:
- make sure that time 10s in future shows up as "next 10s" rather than "next year" (for negative delta we need to iterate the other way)
- if we really want to fix it properly, get server's content's manager time at first load, and include the offset in calculations
